Update documentation for advanced tensorflow usage (#5395)
RasaHQ/rasa: docs/api/tensorflow_usage.rst

Optimizing CPU Performance
--------------------------

.. note::
    It is advisable to configure these options only if you are an advanced tensorflow user and understand
    how different operations are implemented under the hood.

> These options are useful if you have implemented a custom component as part of your pipeline. Based on the tensorflow operations

Reviewer: let's rephrase this to clarify that you should know the workings of an implementation well to know whether these configurations make sense - whether that's their own custom implementation or whether they've figured out our implementation

that the component uses, you can leverage multi-core CPU parallelism by tuning these options.

Parallelizing One Operation
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Set ``TF_INTRA_OP_PARALLELISM_THREADS`` as an environment variable to specify the maximum number of threads that can be used
to parallelize the execution of one operation. For example, operations like ``tf.matmul()`` and ``tf.reduce_sum`` can be executed
on multiple threads running parallely. The default value for this variable is ``0`` which means TensorFlow should

> pick an appropriate value depending on the system configuration.

Reviewer: what does "an appropriate value" actually mean? That it'll run on the maximum amount of possible threads? If yes, why even set this variable to anything?

Author: @akelad TensorFlow also does not mention anything about it in the docs. Assuming it runs on maximum amount of possible threads, as a user this may be undesirable because that hogs my CPU compute capacity and I may want to be conservative in it for other processed.

Author: Okay, I looked at the implementation and default value of 0 makes tensorflow allocate 1 thread per CPU core. Updated this here.

Parallelizing Multiple Operations
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Set ``TF_INTER_OP_PARALLELISM_THREADS`` as an environment variable to specify the maximum number of threads that can be used

> to parallelize the execution of multiple **non-blocking** operations. These would included operations which do not have a

Reviewer: ```suggestion
to parallelize the execution of multiple **non-blocking** operations. These would include operations that do not have a
```

> directed path between them in the tensorflow graph. The default value for this variable is ``0`` which means TensorFlow

Reviewer: directed or direct?

Author: @akelad Should be directed because a direct path is something like A->B between two nodes A and B.
By directed I mean, consider A->B->D and C->D , then there is not directed path between A,C and B,C

Author: Added a small explanation

> should pick an appropriate value depending on the system configuration.

Reviewer: what does appropriate value mean here as well?

Author: Same as above

--------------------------

.. note::
    Since these options affect how operations are carried out under the hood in TensorFlow, it is advisable to
    understand the implementation of different components first and then configure these options. You can always leave
    them to their default value otherwise.

> Depending on the TensorFlow operations an already existing or a custom component uses, you can leverage multi-core CPU

Reviewer: ```suggestion
Depending on the TensorFlow operations a NLU component or Core policy uses, you can leverage multi-core CPU
```

parallelism by tuning these options.

Parallelizing One Operation
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Set ``TF_INTRA_OP_PARALLELISM_THREADS`` as an environment variable to specify the maximum number of threads that can be used
to parallelize the execution of one operation. For example, operations like ``tf.matmul()`` and ``tf.reduce_sum`` can be executed
on multiple threads running in parallel. The default value for this variable is ``0`` which means TensorFlow would
allocate one thread per CPU core.

Parallelizing Multiple Operations
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Set ``TF_INTER_OP_PARALLELISM_THREADS`` as an environment variable to specify the maximum number of threads that can be used
to parallelize the execution of multiple **non-blocking** operations. These would include operations that do not have a
directed path between them in the TensorFlow graph. In other words, the computation of one operation does not affect the

> computation of other operation. The default value for this variable is ``0`` which means TensorFlow would allocate one thread per CPU core.

Reviewer: ```suggestion
computation of the other operation. The default value for this variable is ``0`` which means TensorFlow would allocate one thread per CPU core.
```

^^^^^^^^^^^^^^^^^^^^^^^^^^

TensorFlow by default blocks all the available GPU memory for the running process. This can be limiting if you are running
multiple TensorFlow processes and want to distribute memory across them. To prevent Rasa OS from blocking the

> complete available GPU memory, set the environment variable ``TF_FORCE_GPU_ALLOW_GROWTH`` to ``True``.

Reviewer: ```suggestion
multiple TensorFlow processes and want to distribute memory across them. To prevent Rasa Open Source from blocking all
of the available GPU memory, set the environment variable ``TF_FORCE_GPU_ALLOW_GROWTH`` to ``True``.
```

Restricting Absolute GPU Memory Available
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Often, a developer wants to limit the absolute amount of GPU memory that can be used by a Rasa OS process.

Reviewer: ```suggestion
You may want to limit the absolute amount of GPU memory that can be used by a Rasa Open Source process.
```

> For example, you may have two visible GPUs(``GPU:0`` and ``GPU:1``) and you want to allocate 1024 MB from the first GPU

Reviewer: ```suggestion
For example, say you have two visible GPUs(``GPU:0`` and ``GPU:1``) and you want to allocate 1024 MB from the first GPU
```

> and 2048 MB from the second GPU. You can do so by setting an environment variable as ``TF_GPU_MEMORY_ALLOC="0:1024, 1:2048"``.

Reviewer: ```suggestion
and 2048 MB from the second GPU. You can do this by setting the environment variable  ``TF_GPU_MEMORY_ALLOC`` to ``"0:1024, 1:2048"``.
```

Another scenario can be where you have access to 2 GPUs(``GPU:0`` and ``GPU:1``) but you would like to use only the second

> GPU. ``TF_GPU_MEMORY_ALLOC="1:2048"`` would make 2048 MB of memory available from GPU 1.

Reviewer: do we really think we need the second scenario? Or do we think it's self explanatory with the first?

Author: Agreed, removed.
